File: core/src/main/java/com/sequenceiq/cloudbreak/service/stack/resource/gcp/builders/instance/GcpInstanceResourceBuilder.java
package com.sequenceiq.cloudbreak.service.stack.resource.gcp.builders.instance;

import static com.sequenceiq.cloudbreak.EnvironmentVariableConfig.CB_GCP_AND_AZURE_USER_NAME;
import static com.sequenceiq.cloudbreak.domain.InstanceGroupType.isGateway;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Set;

import javax.inject.Inject;
import javax.inject.Named;

import org.springframework.core.annotation.Order;
import org.springframework.stereotype.Component;

import com.google.api.client.googleapis.json.GoogleJsonResponseException;
import com.google.api.services.compute.Compute;
import com.google.api.services.compute.model.AccessConfig;
import com.google.api.services.compute.model.AttachedDisk;
import com.google.api.services.compute.model.Instance;
import com.google.api.services.compute.model.Metadata;
import com.google.api.services.compute.model.NetworkInterface;
import com.google.api.services.compute.model.Operation;
import com.google.api.services.compute.model.Tags;
import com.google.common.base.Optional;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.Lists;
import com.sequenceiq.cloudbreak.domain.CloudRegion;
import com.sequenceiq.cloudbreak.domain.GcpCredential;
import com.sequenceiq.cloudbreak.domain.GcpDiskMode;
import com.sequenceiq.cloudbreak.domain.GcpDiskType;
import com.sequenceiq.cloudbreak.domain.GcpTemplate;
import com.sequenceiq.cloudbreak.domain.InstanceGroup;
import com.sequenceiq.cloudbreak.domain.InstanceMetaData;
import com.sequenceiq.cloudbreak.domain.Resource;
import com.sequenceiq.cloudbreak.domain.ResourceType;
import com.sequenceiq.cloudbreak.domain.Stack;
import com.sequenceiq.cloudbreak.repository.ClusterRepository;
import com.sequenceiq.cloudbreak.repository.InstanceMetaDataRepository;
import com.sequenceiq.cloudbreak.repository.StackRepository;
import com.sequenceiq.cloudbreak.service.PollingService;
import com.sequenceiq.cloudbreak.service.stack.connector.gcp.GcpRemoveCheckerStatus;
import com.sequenceiq.cloudbreak.service.stack.connector.gcp.GcpRemoveReadyPollerObject;
import com.sequenceiq.cloudbreak.service.stack.connector.gcp.GcpResourceCheckerStatus;
import com.sequenceiq.cloudbreak.service.stack.connector.gcp.GcpResourceException;
import com.sequenceiq.cloudbreak.service.stack.connector.gcp.GcpResourceReadyPollerObject;
import com.sequenceiq.cloudbreak.service.stack.connector.gcp.GcpStackUtil;
import com.sequenceiq.cloudbreak.service.stack.resource.CreateResourceRequest;
import com.sequenceiq.cloudbreak.service.stack.resource.ResourceNameService;
import com.sequenceiq.cloudbreak.service.stack.resource.gcp.GcpSimpleInstanceResourceBuilder;
import com.sequenceiq.cloudbreak.service.stack.resource.gcp.model.GcpDeleteContextObject;
import com.sequenceiq.cloudbreak.service.stack.resource.gcp.model.GcpProvisionContextObject;
import com.sequenceiq.cloudbreak.service.stack.resource.gcp.model.GcpStartStopContextObject;

@Component
@Order(3)
public class GcpInstanceResourceBuilder extends GcpSimpleInstanceResourceBuilder {

    private static final int END_INDEX = 62;
    @Inject
    private StackRepository stackRepository;
    @Inject
    private InstanceMetaDataRepository instanceMetaDataRepository;
    @Inject
    private ClusterRepository clusterRepository;
    @Inject
    private GcpResourceCheckerStatus gcpResourceCheckerStatus;
    @Inject
    private PollingService<GcpResourceReadyPollerObject> gcpInstanceReadyPollerObjectPollingService;
    @Inject
    private GcpRemoveCheckerStatus gcpRemoveCheckerStatus;
    @Inject
    private PollingService<GcpRemoveReadyPollerObject> gcpRemoveReadyPollerObjectPollingService;
    @Inject
    private GcpStackUtil gcpStackUtil;
    @Inject
    @Named("GcpResourceNameService")
    private ResourceNameService resourceNameService;

    @Override
    public Boolean create(final CreateResourceRequest createResourceRequest, String region) throws Exception {
        final GcpInstanceCreateRequest gICR = (GcpInstanceCreateRequest) createResourceRequest;
        Stack stack = stackRepository.findById(gICR.getStackId());
        Compute.Instances.Insert ins =
                gICR.getCompute().instances().insert(gICR.getProjectId(), CloudRegion.valueOf(stack.getRegion()).value(), gICR.getInstance());
        ins.setPrettyPrint(Boolean.TRUE);
        Operation execute = ins.execute();
        if (execute.getHttpErrorStatusCode() == null) {
            Compute.ZoneOperations.Get zoneOperations = createZoneOperations(gICR.getCompute(), gICR.getGcpCredential(), execute,
                    CloudRegion.valueOf(stack.getRegion()));
            GcpResourceReadyPollerObject instReady =
                    new GcpResourceReadyPollerObject(zoneOperations, stack, gICR.getInstance().getName(), execute.getName(), ResourceType.GCP_INSTANCE);
            gcpInstanceReadyPollerObjectPollingService.pollWithTimeout(gcpResourceCheckerStatus, instReady, POLLING_INTERVAL, MAX_POLLING_ATTEMPTS);
            return true;
        } else {
            throw new GcpResourceException(execute.getHttpErrorMessage(), resourceType(), gICR.getInstance().getName());
        }
    }

    private List<AttachedDisk> getAttachedDisks(List<Resource> resources, GcpCredential gcpCredential, CloudRegion zone) {
        List<AttachedDisk> listOfDisks = new ArrayList<>();
        for (Resource resource : filterResourcesByType(resources, ResourceType.GCP_ATTACHED_DISK)) {
            AttachedDisk attachedDisk = new AttachedDisk();
            attachedDisk.setBoot(false);
            attachedDisk.setAutoDelete(true);
            attachedDisk.setType(GcpDiskType.PERSISTENT.getValue());
            attachedDisk.setMode(GcpDiskMode.READ_WRITE.getValue());
            attachedDisk.setDeviceName(resource.getResourceName());
            attachedDisk.setSource(String.format("https://www.googleapis.com/compute/v1/projects/%s/zones/%s/disks/%s",
                    gcpCredential.getProjectId(), zone.value(), resource.getResourceName()));
            listOfDisks.add(attachedDisk);
        }
        return listOfDisks;
    }

    private List<AttachedDisk> getBootDiskList(List<Resource> resources, GcpCredential gcpCredential, CloudRegion zone) {
        List<AttachedDisk> listOfDisks = new ArrayList<>();
        for (Resource resource : filterResourcesByType(resources, ResourceType.GCP_DISK)) {
            AttachedDisk attachedDisk = new AttachedDisk();
            attachedDisk.setBoot(true);
            attachedDisk.setAutoDelete(true);
            attachedDisk.setType(GcpDiskType.PERSISTENT.getValue());
            attachedDisk.setMode(GcpDiskMode.READ_WRITE.getValue());
            attachedDisk.setDeviceName(resource.getResourceName());
            attachedDisk.setSource(String.format("https://www.googleapis.com/compute/v1/projects/%s/zones/%s/disks/%s",
                    gcpCredential.getProjectId(), zone.value(), resource.getResourceName()));
            listOfDisks.add(attachedDisk);
        }
        return listOfDisks;
    }

    @Override
    public Boolean delete(Resource resource, GcpDeleteContextObject deleteContextObject, String region) throws Exception {
        Stack stack = stackRepository.findByIdLazy(deleteContextObject.getStackId());
        try {
            GcpCredential gcpCredential = (GcpCredential) stack.getCredential();
            Operation operation = deleteContextObject.getCompute().instances()
                    .delete(gcpCredential.getProjectId(), CloudRegion.valueOf(region).value(), resource.getResourceName()).execute();
            Compute.ZoneOperations.Get zoneOperations = createZoneOperations(deleteContextObject.getCompute(),
                    gcpCredential, operation, CloudRegion.valueOf(region));
            Compute.GlobalOperations.Get globalOperations = createGlobalOperations(deleteContextObject.getCompute(), gcpCredential, operation);
            GcpRemoveReadyPollerObject gcpRemoveReady =
                    new GcpRemoveReadyPollerObject(zoneOperations, globalOperations, stack, resource.getResourceName(), operation.getName(), resourceType());
            gcpRemoveReadyPollerObjectPollingService.pollWithTimeout(gcpRemoveCheckerStatus, gcpRemoveReady, POLLING_INTERVAL, MAX_POLLING_ATTEMPTS);
        } catch (GoogleJsonResponseException ex) {
            exceptionHandler(ex, resource.getResourceName(), stack);
        } catch (IOException e) {
            throw new GcpResourceException(e);
        }
        return true;
    }

    @Override
    public List<Resource> buildResources(GcpProvisionContextObject provisionContextObject, int index, List<Resource> resources,
            Optional<InstanceGroup> instanceGroup) {
        Stack stack = stackRepository.findById(provisionContextObject.getStackId());
        String resourceName = resourceNameService.resourceName(resourceType(), stack.getName(), instanceGroup.orNull().getGroupName(), index);
        Resource resource = new Resource(resourceType(), resourceName, stack, instanceGroup.orNull().getGroupName());
        return Arrays.asList(resource);
    }

    @Override
    public CreateResourceRequest buildCreateRequest(GcpProvisionContextObject provisionContextObject, List<Resource> resources,
            List<Resource> buildResources, int index, Optional<InstanceGroup> instanceGroup, Optional<String> userData) throws Exception {
        Stack stack = stackRepository.findById(provisionContextObject.getStackId());
        GcpCredential gcpCredential = (GcpCredential) stack.getCredential();
        GcpTemplate gcpTemplate = (GcpTemplate) instanceGroup.orNull().getTemplate();

        List<AttachedDisk> listOfDisks = new ArrayList<>();
        listOfDisks.addAll(getBootDiskList(resources, gcpCredential, CloudRegion.valueOf(stack.getRegion())));
        listOfDisks.addAll(getAttachedDisks(resources, gcpCredential, CloudRegion.valueOf(stack.getRegion())));

        Instance instance = new Instance();
        instance.setMachineType(String.format("https://www.googleapis.com/compute/v1/projects/%s/zones/%s/machineTypes/%s",
                provisionContextObject.getProjectId(), CloudRegion.valueOf(stack.getRegion()).value(), gcpTemplate.getGcpInstanceType().getValue()));
        instance.setName(buildResources.get(0).getResourceName());
        instance.setCanIpForward(Boolean.TRUE);
        instance.setNetworkInterfaces(getNetworkInterface(provisionContextObject, stack.getResources(), CloudRegion.valueOf(stack.getRegion()), instanceGroup));
        instance.setDisks(listOfDisks);
        Tags tags = new Tags();
        tags.setItems(Arrays.asList(instanceGroup.orNull().getGroupName().toLowerCase().toString().replaceAll("[^A-Za-z0-9 ]", "")));
        instance.setTags(tags);
        Metadata metadata = new Metadata();
        metadata.setItems(Lists.<Metadata.Items>newArrayList());

        Metadata.Items sshMetaData = new Metadata.Items();
        sshMetaData.setKey("sshKeys");
        sshMetaData.setValue(CB_GCP_AND_AZURE_USER_NAME + ":" + gcpCredential.getPublicKey());

        Metadata.Items startupScript = new Metadata.Items();
        startupScript.setKey("startup-script");
        startupScript.setValue(userData.orNull());

        metadata.getItems().add(sshMetaData);
        metadata.getItems().add(startupScript);
        instance.setMetadata(metadata);
        return new GcpInstanceCreateRequest(provisionContextObject.getStackId(), resources, instance, provisionContextObject.getProjectId(),
                provisionContextObject.getCompute(), gcpTemplate, gcpCredential, buildResources);
    }

    public Instance describe(Stack stack, Compute compute, Resource resource, CloudRegion region) throws IOException {
        GcpCredential gcpCredential = (GcpCredential) stack.getCredential();
        Compute.Instances.Get getVm = compute.instances().get(gcpCredential.getProjectId(), region.value(),
                resource.getResourceName());
        return getVm.execute();
    }

    private List<NetworkInterface> getNetworkInterface(GcpProvisionContextObject contextObject, Set<Resource> resources, CloudRegion gcpZone,
            Optional<InstanceGroup> instanceGroup) throws IOException {
        NetworkInterface iface = new NetworkInterface();
        String networkName = filterResourcesByType(resources, ResourceType.GCP_NETWORK).get(0).getResourceName();
        iface.setName(networkName);
        AccessConfig accessConfig = new AccessConfig();
        accessConfig.setName(networkName);
        accessConfig.setType("ONE_TO_ONE_NAT");
        if (instanceGroup.isPresent() && isGateway(instanceGroup.orNull().getInstanceGroupType())) {
            Compute.Addresses.Get getReservedIp = contextObject.getCompute().addresses().get(contextObject.getProjectId(), gcpZone.region(),
                    filterResourcesByType(resources, ResourceType.GCP_RESERVED_IP).get(0).getResourceName());
            accessConfig.setNatIP(getReservedIp.execute().getAddress());
        }
        iface.setAccessConfigs(ImmutableList.of(accessConfig));
        iface.setNetwork(String.format("https://www.googleapis.com/compute/v1/projects/%s/global/networks/%s", contextObject.getProjectId(), networkName));
        return Arrays.asList(iface);
    }

    @Override
    public ResourceType resourceType() {
        return ResourceType.GCP_INSTANCE;
    }

    @Override
    public void stop(GcpStartStopContextObject startStopContextObject, Resource resource, String region) {
        GcpCredential credential = (GcpCredential) startStopContextObject.getStack().getCredential();
        try {
            Compute.Instances.Get get = startStopContextObject.getCompute().instances()
                    .get(credential.getProjectId(), CloudRegion.valueOf(region).value(), resource.getResourceName());
            if ("RUNNING".equals(get.execute().getStatus())) {
                Compute.Instances.Stop stop = startStopContextObject.getCompute().instances()
                        .stop(credential.getProjectId(), CloudRegion.valueOf(region).value(), resource.getResourceName());
                stop.setPrettyPrint(Boolean.TRUE);
                setInstanceState(stop.execute(), startStopContextObject, resource, credential, false);
            } else {
                LOGGER.info("Instance is not in RUNNING state - won't stop it.");
            }
        } catch (IOException e) {
            throw new GcpResourceException(String.format("An error occurred while stopping the vm '%s'", resource.getResourceName()), e);
        }
    }

    @Override
    public void start(GcpStartStopContextObject startStopContextObject, Resource resource, String region) {
        GcpCredential credential = (GcpCredential) startStopContextObject.getStack().getCredential();
        try {
            Compute.Instances.Get get = startStopContextObject.getCompute().instances()
                    .get(credential.getProjectId(), CloudRegion.valueOf(region).value(), resource.getResourceName());
            if ("TERMINATED".equals(get.execute().getStatus())) {
                Compute.Instances.Start start = startStopContextObject.getCompute().instances()
                        .start(credential.getProjectId(), CloudRegion.valueOf(region).value(), resource.getResourceName());
                start.setPrettyPrint(Boolean.TRUE);
                setInstanceState(start.execute(), startStopContextObject, resource, credential, true);
            } else {
                LOGGER.info("Instance is not in TERMINATED state - won't start it.");
            }
        } catch (IOException e) {
            throw new GcpResourceException(String.format("An error occurred while starting the vm '%s'", resource.getResourceName()), e);
        }
    }

    private void setInstanceState(Operation operation, GcpStartStopContextObject startStopContextObject, Resource resource,
            GcpCredential credential, boolean start) throws IOException {
        if (operation.getHttpErrorStatusCode() == null) {
            Compute.ZoneOperations.Get zoneOperations = createZoneOperations(
                    startStopContextObject.getCompute(),
                    credential,
                    operation,
                    CloudRegion.valueOf(startStopContextObject.getStack().getRegion()));
            GcpResourceReadyPollerObject instReady = new GcpResourceReadyPollerObject(zoneOperations, startStopContextObject.getStack(),
                    resource.getResourceName(), operation.getName(), ResourceType.GCP_INSTANCE);
            gcpInstanceReadyPollerObjectPollingService.pollWithTimeout(gcpResourceCheckerStatus, instReady, POLLING_INTERVAL, MAX_POLLING_ATTEMPTS);
            if (start) {
                updateInstanceMetadata(startStopContextObject, resource);
            }
        } else {
            throw new GcpResourceException(String.format("Gcp operation returned an error code: %s, message: %s",
                    operation.getHttpErrorStatusCode(), operation.getHttpErrorMessage()));
        }
    }

    private void updateInstanceMetadata(GcpStartStopContextObject startStopContextObject, Resource resource) {
        Instance instance = gcpStackUtil.getInstance(startStopContextObject.getStack(), startStopContextObject.getCompute(), resource);
        if (instance != null) {
            Stack stack = startStopContextObject.getStack();
            InstanceMetaData metaData = instanceMetaDataRepository.findByInstanceId(stack.getId(), resource.getResourceName());
            String publicIP = instance.getNetworkInterfaces().get(0).getAccessConfigs().get(0).getNatIP();
            metaData.setPublicIp(publicIP);
            instanceMetaDataRepository.save(metaData);
        } else {
            LOGGER.error(String.format("Can't find instance by resource name (instance id) : %s", resource.getResourceName()));
        }
    }

    public class GcpInstanceCreateRequest extends CreateResourceRequest {

        private Long stackId;
        private List<Resource> resources = new ArrayList<>();
        private Instance instance;
        private String projectId;
        private Compute compute;
        private GcpTemplate gcpTemplate;
        private GcpCredential gcpCredential;

        public GcpInstanceCreateRequest(Long stackId, List<Resource> resources, Instance instance,
                String projectId, Compute compute, GcpTemplate gcpTemplate, GcpCredential gcpCredential, List<Resource> buildNames) {
            super(buildNames);
            this.stackId = stackId;
            this.resources = resources;
            this.instance = instance;
            this.projectId = projectId;
            this.compute = compute;
            this.gcpTemplate = gcpTemplate;
            this.gcpCredential = gcpCredential;
        }

        public Long getStackId() {
            return stackId;
        }

        public List<Resource> getResources() {
            return resources;
        }

        public Instance getInstance() {
            return instance;
        }

        public String getProjectId() {
            return projectId;
        }

        public Compute getCompute() {
            return compute;
        }

        public GcpTemplate getGcpTemplate() {
            return gcpTemplate;
        }

        public GcpCredential getGcpCredential() {
            return gcpCredential;
        }
    }

}

File: cloud-gcp/src/main/java/com/sequenceiq/cloudbreak/cloud/gcp/compute/GcpInstanceResourceBuilder.kt
package com.sequenceiq.cloudbreak.cloud.gcp.compute

import java.util.Arrays.asList

import java.io.IOException
import java.util.ArrayList
import java.util.Arrays

import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component

import com.google.api.client.googleapis.json.GoogleJsonResponseException
import com.google.api.services.compute.Compute
import com.google.api.services.compute.model.AccessConfig
import com.google.api.services.compute.model.AttachedDisk
import com.google.api.services.compute.model.Instance
import com.google.api.services.compute.model.Metadata
import com.google.api.services.compute.model.NetworkInterface
import com.google.api.services.compute.model.Operation
import com.google.api.services.compute.model.Tags
import com.sequenceiq.cloudbreak.api.model.InstanceGroupType
import com.sequenceiq.cloudbreak.cloud.context.AuthenticatedContext
import com.sequenceiq.cloudbreak.cloud.context.CloudContext
import com.sequenceiq.cloudbreak.cloud.gcp.GcpResourceException
import com.sequenceiq.cloudbreak.cloud.gcp.context.GcpContext
import com.sequenceiq.cloudbreak.cloud.gcp.util.GcpStackUtil
import com.sequenceiq.cloudbreak.cloud.model.AvailabilityZone
import com.sequenceiq.cloudbreak.cloud.model.CloudInstance
import com.sequenceiq.cloudbreak.cloud.model.CloudResource
import com.sequenceiq.cloudbreak.cloud.model.CloudVmInstanceStatus
import com.sequenceiq.cloudbreak.cloud.model.Group
import com.sequenceiq.cloudbreak.cloud.model.Image
import com.sequenceiq.cloudbreak.cloud.model.InstanceStatus
import com.sequenceiq.cloudbreak.cloud.model.InstanceTemplate
import com.sequenceiq.cloudbreak.cloud.model.Location
import com.sequenceiq.cloudbreak.cloud.model.Region
import com.sequenceiq.cloudbreak.common.type.ResourceType

@Component
class GcpInstanceResourceBuilder : AbstractGcpComputeBuilder() {

    override fun create(context: GcpContext, privateId: Long, auth: AuthenticatedContext, group: Group, image: Image): List<CloudResource> {
        val cloudContext = auth.cloudContext
        val resourceName = resourceNameService.resourceName(resourceType(), cloudContext.name, group.name, privateId)
        return Arrays.asList(createNamedResource(resourceType(), resourceName))
    }

    @Throws(Exception::class)
    override fun build(context: GcpContext, privateId: Long, auth: AuthenticatedContext, group: Group, image: Image,
                       buildableResource: List<CloudResource>): List<CloudResource> {
        val template = group.instances[0].template
        val projectId = context.projectId
        val location = context.location

        val compute = context.compute

        val computeResources = context.getComputeResources(privateId)
        val listOfDisks = ArrayList<AttachedDisk>()
        listOfDisks.addAll(getBootDiskList(computeResources, projectId, location.availabilityZone))
        listOfDisks.addAll(getAttachedDisks(computeResources, projectId, location.availabilityZone))

        val instance = Instance()
        instance.machineType = String.format("https://www.googleapis.com/compute/v1/projects/%s/zones/%s/machineTypes/%s",
                projectId, location.availabilityZone.value(), template.flavor)
        instance.name = buildableResource[0].name
        instance.canIpForward = java.lang.Boolean.TRUE
        instance.networkInterfaces = getNetworkInterface(context.networkResources, location.region, group, compute, projectId)
        instance.disks = listOfDisks

        val tags = Tags()
        val tagList = ArrayList<String>()
        tagList.add(group.name.toLowerCase().replace("[^A-Za-z0-9 ]".toRegex(), ""))
        tagList.add(GcpStackUtil.getClusterTag(auth.cloudContext))
        tags.items = tagList
        instance.tags = tags

        val metadata = Metadata()
        metadata.items = ArrayList<Metadata.Items>()

        val sshMetaData = Metadata.Items()
        sshMetaData.key = "sshKeys"
        sshMetaData.value = auth.cloudCredential.loginUserName + ":" + auth.cloudCredential.publicKey

        val startupScript = Metadata.Items()
        startupScript.key = "startup-script"
        startupScript.value = image.getUserData(group.type)

        metadata.items.add(sshMetaData)
        metadata.items.add(startupScript)
        instance.metadata = metadata

        val insert = compute.instances().insert(projectId, location.availabilityZone.value(), instance)
        insert.prettyPrint = java.lang.Boolean.TRUE
        try {
            val operation = insert.execute()
            if (operation.httpErrorStatusCode != null) {
                throw GcpResourceException(operation.httpErrorMessage, resourceType(), buildableResource[0].name)
            }
            return asList(createOperationAwareCloudResource(buildableResource[0], operation))
        } catch (e: GoogleJsonResponseException) {
            throw GcpResourceException(checkException(e), resourceType(), buildableResource[0].name)
        }

    }

    @Throws(Exception::class)
    override fun delete(context: GcpContext, auth: AuthenticatedContext, resource: CloudResource): CloudResource? {
        val resourceName = resource.name
        try {
            val operation = context.compute.instances().delete(context.projectId, context.location.availabilityZone.value(), resourceName).execute()
            return createOperationAwareCloudResource(resource, operation)
        } catch (e: GoogleJsonResponseException) {
            exceptionHandler(e, resourceName, resourceType())
        }

        return null
    }

    override fun checkInstances(context: GcpContext, auth: AuthenticatedContext, instances: List<CloudInstance>): List<CloudVmInstanceStatus>? {
        val cloudInstance = instances[0]
        try {
            LOGGER.info("Checking instance: {}", cloudInstance)
            val operation = check(context, cloudInstance)
            val finished = operation == null || GcpStackUtil.analyzeOperation(operation)
            val status = if (finished) if (context.isBuild) InstanceStatus.STARTED else InstanceStatus.STOPPED else InstanceStatus.IN_PROGRESS
            LOGGER.info("Instance: {} status: {}", instances, status)
            return asList(CloudVmInstanceStatus(cloudInstance, status))
        } catch (e: Exception) {
            LOGGER.info("Failed to check instance state of {}", cloudInstance)
            return asList(CloudVmInstanceStatus(cloudInstance, InstanceStatus.IN_PROGRESS))
        }

    }

    override fun stop(context: GcpContext, auth: AuthenticatedContext, instance: CloudInstance): CloudVmInstanceStatus? {
        return stopStart(context, auth, instance, true)
    }

    override fun start(context: GcpContext, auth: AuthenticatedContext, instance: CloudInstance): CloudVmInstanceStatus? {
        return stopStart(context, auth, instance, false)
    }

    override fun resourceType(): ResourceType {
        return ResourceType.GCP_INSTANCE
    }

    override fun order(): Int {
        return 2
    }

    private fun getBootDiskList(resources: List<CloudResource>, projectId: String, zone: AvailabilityZone): List<AttachedDisk> {
        val listOfDisks = ArrayList<AttachedDisk>()
        for (resource in filterResourcesByType(resources, ResourceType.GCP_DISK)) {
            listOfDisks.add(createDisk(resource, projectId, zone, true))
        }
        return listOfDisks
    }

    private fun getAttachedDisks(resources: List<CloudResource>, projectId: String, zone: AvailabilityZone): List<AttachedDisk> {
        val listOfDisks = ArrayList<AttachedDisk>()
        for (resource in filterResourcesByType(resources, ResourceType.GCP_ATTACHED_DISK)) {
            listOfDisks.add(createDisk(resource, projectId, zone, false))
        }
        return listOfDisks
    }

    private fun createDisk(resource: CloudResource, projectId: String, zone: AvailabilityZone, boot: Boolean): AttachedDisk {
        val attachedDisk = AttachedDisk()
        attachedDisk.boot = boot
        attachedDisk.autoDelete = true
        attachedDisk.type = GCP_DISK_TYPE
        attachedDisk.mode = GCP_DISK_MODE
        attachedDisk.deviceName = resource.name
        attachedDisk.source = String.format("https://www.googleapis.com/compute/v1/projects/%s/zones/%s/disks/%s",
                projectId, zone.value(), resource.name)
        return attachedDisk
    }

    @Throws(IOException::class)
    private fun getNetworkInterface(resources: List<CloudResource>,
                                    region: Region, group: Group, compute: Compute, projectId: String): List<NetworkInterface> {
        val networkInterface = NetworkInterface()
        val subnet = filterResourcesByType(resources, ResourceType.GCP_SUBNET)
        val networkName = if (subnet.isEmpty()) filterResourcesByType(resources, ResourceType.GCP_NETWORK)[0].name else subnet[0].name
        networkInterface.name = networkName
        val accessConfig = AccessConfig()
        accessConfig.name = networkName
        accessConfig.type = "ONE_TO_ONE_NAT"
        if (InstanceGroupType.GATEWAY == group.type) {
            val getReservedIp = compute.addresses().get(projectId, region.value(),
                    filterResourcesByType(resources, ResourceType.GCP_RESERVED_IP)[0].name)
            accessConfig.natIP = getReservedIp.execute().address
        }
        networkInterface.accessConfigs = asList(accessConfig)
        if (subnet.isEmpty()) {
            networkInterface.network = String.format("https://www.googleapis.com/compute/v1/projects/%s/global/networks/%s", projectId, networkName)
        } else {
            networkInterface.subnetwork = String.format("https://www.googleapis.com/compute/v1/projects/%s/regions/%s/subnetworks/%s", projectId, region.value(), networkName)
        }
        return asList(networkInterface)
    }

    private fun filterResourcesByType(resources: Collection<CloudResource>, resourceType: ResourceType): List<CloudResource> {
        val resourcesTemp = ArrayList<CloudResource>()
        for (resource in resources) {
            if (resourceType == resource.type) {
                resourcesTemp.add(resource)
            }
        }
        return resourcesTemp
    }

    private fun stopStart(context: GcpContext, auth: AuthenticatedContext, instance: CloudInstance, stopRequest: Boolean): CloudVmInstanceStatus? {
        val projectId = GcpStackUtil.getProjectId(auth.cloudCredential)
        val availabilityZone = context.location.availabilityZone.value()
        val compute = context.compute
        val instanceId = instance.instanceId
        try {
            val get = compute.instances().get(projectId, availabilityZone, instanceId)
            val state = if (stopRequest) "RUNNING" else "TERMINATED"
            if (state == get.execute().status) {
                val operation = if (stopRequest)
                    compute.instances().stop(projectId, availabilityZone, instanceId).setPrettyPrint(true).execute()
                else
                    compute.instances().start(projectId, availabilityZone, instanceId).setPrettyPrint(true).execute()
                val operationAwareCloudInstance = createOperationAwareCloudInstance(instance, operation)
                return checkInstances(context, auth, asList(operationAwareCloudInstance))!![0]
            } else {
                LOGGER.info("Instance {} is not in {} state - won't start it.", state, instanceId)
                return null
            }
        } catch (e: IOException) {
            throw GcpResourceException(String.format("An error occurred while stopping the vm '%s'", instanceId), e)
        }

    }

    companion object {

        private val LOGGER = LoggerFactory.getLogger(GcpInstanceResourceBuilder::class.java)
        private val GCP_DISK_TYPE = "PERSISTENT"
        private val GCP_DISK_MODE = "READ_WRITE"
    }

}
